File: atmos/tables.f90
MODULE PhysicalConstants
  IMPLICIT NONE

  REAL,PARAMETER:: REARTH = 6369.0                 ! radius of the Earth (km)
  REAL,PARAMETER:: GMR = 34.163195                             ! gas constant

  REAL,PARAMETER:: FT2METERS = 0.3048               ! mult. ft. to get meters (exact)
  REAL,PARAMETER:: KELVIN2RANKINE = 1.8             ! mult deg K to get deg R
  REAL,PARAMETER:: PSF2NSM = 47.880258          ! mult lb/sq.ft to get N/sq.m
  REAL,PARAMETER:: SCF2KCM = 515.379         ! mult slug/cu.ft to get kg/cu.m
  REAL,PARAMETER:: TZERO = 288.15            ! temperature at sealevel, deg K
  REAL,PARAMETER:: PZERO = 101325.0            ! pressure at sealevel, N/sq.m
  REAL,PARAMETER:: RHOZERO = 1.2250            ! density at sealevel, kg/cu.m
  REAL,PARAMETER:: ASOUNDZERO = 340.294   ! speed of sound at sealevel, m/sec

  REAL,PARAMETER:: BETAVISC = 1.458E-6 ! viscosity term, N sec/(sq.m sqrt(deg K)
  REAL,PARAMETER:: SUTH = 110.4              ! Sutherland's constant, deg K

END MODULE PhysicalConstants   ! --------------------------------------------

!+
PROGRAM Tables                                         ! \atmtable\tables.f90
! ---------------------------------------------------------------------------
! PURPOSE - Make tables of atmospheric properties
!
! AUTHOR - Ralph L. Carmichael, Public Domain Aeronautical Software
!
!     REVISION HISTORY
!   DATE  VERS PERSON  STATEMENT OF CHANGES
! 28Feb95  1.0   RLC   Assembled several old codes
!  7Jul95  1.1   RLC   Renamed routines, coded viscosity in US units
! 26Jul95  1.2   RLC   Converted f77 program to f90 style.
!                      Made Module PhysicalConstants
!  6Aug95  1.3   RLC   Replaced 1962 tables with 1976 tables
! 25Aug95  1.4   RLC   Added MODIFIER
! 19Sep95  1.5   RLC   Added a little precision to pressure tables
!  3Oct95  1.6   RLC   Numerous style changes to run under Elf90
! 29Nov96  1.7   RLC   Error message if unable to open output file
! 29Oct01  1.8   RLC   Renamed output files
!
! NOTES-

  IMPLICIT NONE
!============================================================================
!     C O N S T A N T S                                                     |
!============================================================================
  CHARACTER(LEN=*),PARAMETER::GREETING =  &
        ' tables - A Fortran 95 program to make atmosphere tables'
  CHARACTER(LEN=*),PARAMETER::AUTHOR=  &
    ' Ralph L. Carmichael, Public Domain Aeronautical Software'
  CHARACTER(LEN=*),PARAMETER::MODIFIER=' '   ! put your name here if you mod it
  CHARACTER(LEN=*),PARAMETER::VERSION=' 1.8 (29Oct01)'
  CHARACTER(LEN=*),PARAMETER::FAREWELL= &
    'Four files have been added to your directory.'
!-----------------------------------------------------------------------
  WRITE(*,*) GREETING
  WRITE(*,*) AUTHOR
  IF (MODIFIER /= ' ') WRITE(*,*) 'Modified by '//MODIFIER
  WRITE(*,*) VERSION

  CALL LongUSTable()
  CALL ShortUSTable()
  CALL LongSItable()
  CALL ShortSItable()

  WRITE(*,*) FAREWELL
  WRITE(*,*) 'tables has terminated successfully.'
  STOP

CONTAINS

!+
SUBROUTINE LongUSTable()
! ---------------------------------------------------------------------------
  USE PhysicalConstants

  CHARACTER(LEN=*),PARAMETER::HEAD1=' alt   sigma    delta   theta  temp'// &
      '   press    dens      a    visc  k.visc'
  CHARACTER(LEN=*),PARAMETER::HEAD2=' Kft                           degR'// &
      ' lb/sq.ft s/cu.ft    fps s/ft-s sq.ft/s'

  INTEGER,PARAMETER::ITXT=1
  CHARACTER(LEN=*),PARAMETER:: FMT = &
    '(I4,2ES9.3E1,F7.4,F6.1,2ES9.3E1,F7.1,F6.3,ES8.2E1)'

  INTEGER:: errCode
  INTEGER:: i                              ! altitude in thousand feet
  REAL:: altKm                                       !  altitude in Km
  REAL:: sigma,delta,theta       ! density,pressure,temperature ratios
  REAL:: temp,pressure,density
  REAL:: asound                             ! speed of sound in ft/sec
  REAL:: viscosity                       ! viscosity in slugs/(ft*sec)
  REAL:: kinematicViscosity                                ! sq.ft/sec
!----------------------------------------------------------------------------
  OPEN(UNIT=ITXT, FILE='us1f.prt', &
     IOSTAT=errCode, STATUS='REPLACE', ACTION='WRITE')
  IF (errCode /= 0) THEN
    WRITE(*,*) "Unable to open output file us1f.prt"
    STOP
  END IF
  WRITE(ITXT,*) HEAD1
  WRITE(ITXT,*) HEAD2

  DO i=-1,56   ! vary alt (ft) from -5000 to 280,000 by 5000
    altKm=REAL(5*i)*FT2METERS
    CALL Atmosphere(altKm, sigma,delta,theta)
    temp=KELVIN2RANKINE*TZERO*theta
    pressure=(PZERO/PSF2NSM)*delta
    density=(RHOZERO/SCF2KCM)*sigma
    asound=(ASOUNDZERO/FT2METERS)*Sqrt(theta)
    viscosity=(1.0/PSF2NSM)*MetricViscosity(theta)
    kinematicViscosity=viscosity/density
    WRITE(ITXT, FMT) 5*i, sigma, delta, theta, temp, pressure, &
      density, asound, 1E6*viscosity, kinematicViscosity
  END DO
  Close(UNIT=ITXT)
  RETURN
 
END Subroutine LongUStable   ! ----------------------------------------------

!+
SUBROUTINE ShortUStable()
!   --------------------------------------------------------------------
  USE PhysicalConstants

! L O C A L   C O N S T A N T S
  CHARACTER(LEN=*),PARAMETER::HEAD1=' alt  sigma  delta  theta  temp'//  &
      '  press    dens     a    visc  k.visc  ratio'
  CHARACTER(LEN=*),PARAMETER::HEAD2=' Kft                       degR   psf  '//  &
      ' s/cu.ft   fps s/ft-sec sq.ft/s 1/ft'
  CHARACTER(LEN=*),PARAMETER:: FMT = &
      '(I4,3F7.4,F6.1,F7.1,F10.7,F7.1,F6.3,ES8.2E1,F6.2)'
  INTEGER,PARAMETER::ITXT=1

! L O C A L   V A R I A B L E S
  INTEGER:: errCode
  INTEGER:: i !     altitude in thousand feet
  REAL:: altKm  !
  REAL:: sigma,delta,theta
  REAL:: temp,pressure,density,asound
  REAL:: viscosity,kinematicViscosity,vratio
!----------------------------------------------------------------------------
  OPEN(UNIT=ITXT, FILE='us2f.prt', &
    IOSTAT=errCode, STATUS='REPLACE', ACTION='WRITE')
  IF (errCode /= 0) THEN
    WRITE(*,*) "Unable to open output file us2f.prt"
    STOP
  END IF
  WRITE(ITXT,*) HEAD1
  WRITE(ITXT,*) HEAD2
  DO i=-1,65   ! vary alt (ft) from -1000 to 65000 by 1000
    altKm=REAL(i)*FT2METERS
    CALL SimpleAtmosphere(altKm, sigma,delta,theta)
    temp=(KELVIN2RANKINE*TZERO)*theta
    pressure=(PZERO/PSF2NSM)*delta
    density=(RHOZERO/SCF2KCM)*sigma
    asound=(ASOUNDZERO/FT2METERS)*Sqrt(theta)
    viscosity=(1.0/PSF2NSM)*MetricViscosity(theta)
    kinematicViscosity=viscosity/density
    vratio=1.0E-6*asound/kinematicViscosity
    WRITE(ITXT,FMT) i, sigma, delta, theta, temp, pressure,    &
      density, asound, 1E6*viscosity, kinematicViscosity, vratio
  END DO
  Close(Itxt)
  RETURN
END Subroutine ShortUSTable   ! ---------------------------------------------

!+
SUBROUTINE LongSItable()
! ---------------------------------------------------------------------------
  USE PhysicalConstants

! L O C A L   C O N S T A N T S
  CHARACTER(LEN=*),PARAMETER::HEAD1=' alt    sigma     delta   theta  temp'// &
    '   press    dens     a    visc  k.visc'
  CHARACTER(LEN=*),PARAMETER::HEAD2='  Km                             degK'// &
    '  N/sq.m   kg/cu.m m/sec kg/m-s sq.m/s'
  CHARACTER(LEN=*),PARAMETER:: FMT = &
    '(I4,2ES10.4E1,F7.4,F6.1,2ES9.3E1,F6.1,F6.2,2ES8.2E1)'
  INTEGER,PARAMETER::ITXT=1

! L O C A L   V A R I A B L E S
  INTEGER:: errCode
  INTEGER:: i
  REAL:: altKm                                ! altitude in kilometers
  REAL:: sigma,delta,theta
  REAL:: temp,pressure,density, asound
  REAL:: viscosity,kinematicViscosity
!----------------------------------------------------------------------------
  OPEN(UNIT=ITXT, FILE='si1f.prt', &
    IOSTAT=errCode, STATUS='REPLACE', ACTION='WRITE')
  IF (errCode /= 0) THEN
    WRITE(*,*) "Unable to open output file si1f.prt"
    STOP
  END IF
  WRITE(ITXT,*) HEAD1
  WRITE(ITXT,*) HEAD2

  DO i=-2,86,2
    altKm=REAL(i)
    CALL Atmosphere(altKm, sigma,delta,theta)
    temp=TZERO*theta
    pressure=PZERO*delta
    density=RHOZERO*sigma
    asound=ASOUNDZERO*Sqrt(theta)
    viscosity=MetricViscosity(theta)
    kinematicViscosity=viscosity/density
    WRITE(ITXT,FMT) i, sigma, delta, theta, temp, pressure,  &
      density, asound, 1.0E6*viscosity, kinematicViscosity
  END DO
  Close(Itxt)
  RETURN
END Subroutine LongSItable   ! ----------------------------------------------

!+
SUBROUTINE ShortSItable()
! ---------------------------------------------------------------------------
  USE PhysicalConstants

! L O C A L   C O N S T A N T S
  CHARACTER(LEN=*),PARAMETER::HEAD1=' alt  sigma  delta  theta  temp'//   &
    '  press  dens   a    visc  k.visc ratio'
  CHARACTER(LEN=*),PARAMETER::HEAD2='  Km                       degK N/sq.m'// &
    '  kcm  m/sec kg/m-s sq.m/s  1/m'
  CHARACTER(LEN=*),PARAMETER:: FMT = &
    '(F4.1,3F7.4,F6.1,I7,F6.3,F6.1,F6.2,ES8.2E1,F6.2)'
  INTEGER,PARAMETER::ITXT=1                   ! unit number of output file

! L O C A L   V A R I A B L E S
  INTEGER:: errCode
  INTEGER:: i                                   !  steps thru altitude
  REAL:: altKm
  REAL:: sigma,delta,theta
  REAL:: temp,pressure,density, asound
  REAL:: viscosity,kinematicViscosity,vratio
!----------------------------------------------------------------------------
  OPEN(UNIT=ITXT, FILE='si2f.prt', &
    IOSTAT=errCode, STATUS='REPLACE', ACTION='WRITE')
  IF (errCode /= 0) THEN
    WRITE(*,*) "Unable to open output file si2f.prt"
    STOP
  END IF
  WRITE(ITXT,*) HEAD1
  WRITE(ITXT,*) HEAD2

  DO i=-1,40
    altKm=0.5*REAL(i)
    CALL SimpleAtmosphere(altKm, sigma,delta,theta)
    temp=TZERO*theta
    pressure=PZERO*delta
    density=RHOZERO*sigma
    asound=ASOUNDZERO*Sqrt(theta)
    viscosity=MetricViscosity(theta)
    kinematicViscosity=viscosity/density
    vratio=asound/kinematicViscosity
    WRITE(ITXT,FMT) altKm, sigma, delta, theta, temp, INT(pressure),  &
      density, asound, 1.0E6*viscosity, kinematicViscosity, 1.0E-6*vratio
  END DO
  Close(Itxt)
  RETURN
 
END Subroutine ShortSITable   ! ---------------------------------------------

!+
SUBROUTINE Atmosphere(alt, sigma, delta, theta)
!   -------------------------------------------------------------------------
! PURPOSE - Compute the properties of the 1976 standard atmosphere to 86 km.

  IMPLICIT NONE
!============================================================================
!     A R G U M E N T S                                                     |
!============================================================================
  REAL,INTENT(IN)::  alt    ! geometric altitude, km.                        
  REAL,INTENT(OUT):: sigma  ! density/sea-level standard density              
  REAL,INTENT(OUT):: delta  ! pressure/sea-level standard pressure           
  REAL,INTENT(OUT):: theta  ! temperature/sea-level standard temperature
!============================================================================
!     L O C A L   C O N S T A N T S                                         |
!============================================================================
  REAL,PARAMETER:: REARTH = 6369.0                 ! radius of the Earth (km)
  REAL,PARAMETER:: GMR = 34.163195                             ! gas constant
  INTEGER,PARAMETER:: NTAB=8       ! number of entries in the defining tables
!============================================================================
!     L O C A L   V A R I A B L E S                                         |
!============================================================================
  INTEGER:: i,j,k                                                  ! counters
  REAL:: h                                       ! geopotential altitude (km)
  REAL:: tgrad, tbase      ! temperature gradient and base temp of this layer
  REAL:: tlocal                                           ! local temperature
  REAL:: deltah                             ! height above base of this layer
!============================================================================
!     L O C A L   A R R A Y S   ( 1 9 7 6   S T D.  A T M O S P H E R E )   |
!============================================================================
  REAL,DIMENSION(NTAB),PARAMETER:: htab= &
                          (/0.0, 11.0, 20.0, 32.0, 47.0, 51.0, 71.0, 84.852/)
  REAL,DIMENSION(NTAB),PARAMETER:: ttab= &
          (/288.15, 216.65, 216.65, 228.65, 270.65, 270.65, 214.65, 186.946/)
  REAL,DIMENSION(NTAB),PARAMETER:: ptab= &
               (/1.0, 2.233611E-1, 5.403295E-2, 8.5666784E-3, 1.0945601E-3, &
                                     6.6063531E-4, 3.9046834E-5, 3.68501E-6/)
  REAL,DIMENSION(NTAB),PARAMETER:: gtab= &
                                (/-6.5, 0.0, 1.0, 2.8, 0.0, -2.8, -2.0, 0.0/)
!----------------------------------------------------------------------------
  h=alt*REARTH/(alt+REARTH)      ! convert geometric to geopotential altitude

  i=1 
  j=NTAB                                  ! setting up for binary search
  DO
    k=(i+j)/2                                              ! integer division
    IF (h < htab(k)) THEN
      j=k
    ELSE
      i=k
    END IF   
    IF (j <= i+1) EXIT
  END DO

  tgrad=gtab(i)                                     ! i will be in 1...NTAB-1
  tbase=ttab(i)
  deltah=h-htab(i)
  tlocal=tbase+tgrad*deltah
  theta=tlocal/ttab(1)                                    ! temperature ratio

  IF (tgrad == 0.0) THEN                                     ! pressure ratio
    delta=ptab(i)*EXP(-GMR*deltah/tbase)
  ELSE
    delta=ptab(i)*(tbase/tlocal)**(GMR/tgrad)
  END IF

  sigma=delta/theta                                           ! density ratio
  RETURN
END Subroutine Atmosphere   ! -----------------------------------------------

!+
SUBROUTINE SimpleAtmosphere(alt,sigma,delta,theta)
!   -------------------------------------------------------------------------
! PURPOSE - Compute the characteristics of the lower atmosphere.

! NOTES-Correct to 20 km. Only approximate above there

  IMPLICIT NONE
!============================================================================
!     A R G U M E N T S                                                     |
!============================================================================
  REAL,INTENT(IN)::  alt    ! geometric altitude, km.
  REAL,INTENT(OUT):: sigma  ! density/sea-level standard density             
  REAL,INTENT(OUT):: delta  ! pressure/sea-level standard pressure            
  REAL,INTENT(OUT):: theta  ! temperature/sea-level standard temperature   
!============================================================================
!     L O C A L   C O N S T A N T S                                         |
!============================================================================
  REAL,PARAMETER:: REARTH = 6369.0                ! radius of the Earth (km)
  REAL,PARAMETER:: GMR = 34.163195                            ! gas constant
!============================================================================
!     L O C A L   V A R I A B L E S                                         |
!============================================================================
  REAL:: h   ! geopotential altitude
!----------------------------------------------------------------------------
  h=alt*REARTH/(alt+REARTH)      ! convert geometric to geopotential altitude

  IF (h < 11.0) THEN
    theta=1.0+(-6.5/288.15)*h                                   ! Troposphere
    delta=theta**(GMR/6.5)
  ELSE
    theta=216.65/288.15                                        ! Stratosphere
    delta=0.2233611*EXP(-GMR*(h-11.0)/216.65)
  END IF

  sigma=delta/theta
  RETURN
END Subroutine SimpleAtmosphere   ! -----------------------------------------

!+
FUNCTION MetricViscosity(theta) RESULT(visc)
!   -------------------------------------------------------------------------
! PURPOSE - Compute viscosity using Sutherland's formula.
!        Returns viscosity in kg/(meter-sec)
  USE PhysicalConstants

  IMPLICIT NONE
!============================================================================
!     A R G U M E N T S                                                     |
!============================================================================
  REAL,INTENT(IN) :: theta                ! temperature/sea-level temperature  
!============================================================================
!     L O C A L   V A R I A B L E S                                         |
!============================================================================
  REAL:: visc
  REAL:: temp                              ! temperature in deg Kelvin
!----------------------------------------------------------------------------
  temp=TZERO*theta
  visc=BETAVISC*Sqrt(temp*temp*temp)/(temp+SUTH)
  RETURN
END Function MetricViscosity   ! --------------------------------------------

END Program Tables   ! ------------------------------------------------------

File: atmos/tables.cpp
// PROGRAM Tables;                                    { \atmtable\Tables.cpp
// --------------------------------------------------------------------------
// PURPOSE - Make tables of atmospheric properties
// AUTHOR  - Ralph L. Carmichael, Public Domain Aeronautical Software
// REFERENCE - U.S. Standard Atmosphere, 1976. U.S. Govt Printing Office
// REVISION HISTORY
//   DATE  VERS PERSON  STATEMENT OF CHANGES
// 26Feb95  1.0   RLC   Assembled several old codes
//  7Jul95  1.1   RLC   Added viscosity calculations
//  6Aug95  1.2   RLC   Replaced 1962 tables with 1976
// 25Aug95  1.3   RLC   Added MODIFIER
// 19Sep95  1.4   RLC   Added a little precision to pressure table
// 31Oct01  1.5   RLC   Renamed the output files
// 31Jul08  1.6   RLC   Adapted to standard library namespace
// --------------------------------------------------------------------------
#include <iostream>
#include <fstream>
#include <iomanip>
#include <cmath>
#include <string>
using namespace std;
//#include <string.h>

  const char* const VERSION = "1.6 (31 Jul 2008)";
  const char* const GREETING =
	  "Tables - A C++ program to compute atmosphere tables";
  const char* const AUTHOR =
	  "Ralph L. Carmichael, Public Domain Aeronautical Software";
//  const char* const MODIFIER = "";
  string MODIFIER ("");
  const char* const FAREWELL = "Four files added to your directory.";
  const char* const FINALMESS =
	  "Normal termination of tables, version ";

// P H Y S I C A L   C O N S T A N T S
  const double FT2METERS      = 0.3048;     // mult. ft. to get meters (exact)
  const double KELVIN2RANKINE = 1.8;              // mult kelvins to get deg R
  const double PSF2NSM        = 47.880258;      // mult lb/sq.ft to get N/sq.m
  const double SCF2KCM        = 515.379;    // mult slugs/cu.ft to get kg/cu.m
  const double TZERO          = 288.15;      // sea level temperature, kelvins
  const double PZERO          = 101325.0;        // sea-level pressure, N/sq.m
  const double RHOZERO        = 1.225;           // sea level density, kg/cu.m
  const double AZERO          = 340.294;    // sea-level speed of sound, m/sec

// F U N C T I O N   P R O T O T Y P E S
  void SimpleAtmosphere(const double, double&, double&, double&);
  void Atmosphere(const double, double&, double&, double&);
  void LongUSTable(void);
  void ShortUSTable(void);
  void LongSITable(void);
  void ShortSITable(void);
  double MetricViscosity(const double);

// ==========================================================================
int main(int argc, char* argv[])
{
  std::cout << "Executing " << argv[0] << '\n';
  std::cout << GREETING << '\n' << AUTHOR << '\n';
//  if (strlen(MODIFIER) > 0) std::cout << "Modified by " << MODIFIER << '\n';
  if (MODIFIER.length() > 0) std::cout << "Modified by " << MODIFIER << '\n';

  std::cout << "     version " << VERSION << std::endl;
  LongUSTable();
  ShortUSTable();
  LongSITable();
  ShortSITable();
  std::cout << FAREWELL << '\n' << FINALMESS << VERSION << std::endl;
  return 0;
}   // ----------------------------------------------- End of Program Tables.

// ==========================================================================
void LongUSTable(void)
{
  double altKm;
  double sigma,delta,theta;
  double temp,pressure,density,asound;
  double viscosity,kinematicViscosity;

  std::ofstream Itxt("us1cpp.prt");
  Itxt << "units of alt are Kft; pressure lbs/sq.ft.; density is slugs/cu.ft.\n";
  Itxt << "speed of sound is ft/second; viscosity is slugs per ft-second;\n"; 
  Itxt << "kinematic viscosity is square feet per second.\n";
  Itxt << " alt     sigma      delta   theta ";
  Itxt << " temp    press      dens      a     visc   k.visc" << std::endl;
//  Itxt << " Kft                              ";
//  Itxt << " degR  lb/sq.ft   sl/cu.ft   fps s/ft-s sq.ft/s\n";
  Itxt.setf(std::ios::showpoint | std::ios::uppercase);

  for(int i=-1; i<=56; i++)  // 5*i is altitude in thousand feet }
    {
      altKm=5*i*FT2METERS;
      Atmosphere(altKm, sigma,delta,theta);
      Itxt.width(4); Itxt << 5*i;
      Itxt.setf(std::ios::scientific, std::ios::floatfield);
      Itxt.width(11); Itxt.precision(3); Itxt << sigma;
      Itxt.width(11); Itxt << delta;
      Itxt.setf(std::ios::fixed, std::ios::floatfield);
      Itxt.width(7); Itxt.precision(4); Itxt << theta;

      temp=(KELVIN2RANKINE*TZERO)*theta;
      pressure=(PZERO/PSF2NSM)*delta;
      density=(RHOZERO/SCF2KCM)*sigma;
      asound=(AZERO/FT2METERS)*sqrt(theta);

      Itxt.width(6); Itxt.precision(1); Itxt << temp;
      Itxt.setf(std::ios::scientific, std::ios::floatfield);
      Itxt.width(11); Itxt.precision(3); Itxt << pressure;
      Itxt.width(11); Itxt.precision(3); Itxt << density;
      Itxt.setf(std::ios::fixed, std::ios::floatfield);
      Itxt.width(7); Itxt.precision(1); Itxt << asound;

      viscosity=(1.0/PSF2NSM)*MetricViscosity(theta);
      kinematicViscosity=viscosity/density;
      Itxt.width(6); Itxt.precision(3); Itxt << 1.0E6*viscosity;
      Itxt.setf(std::ios::scientific, std::ios::floatfield);
      Itxt.width(10); Itxt.precision(2); Itxt << kinematicViscosity << std::endl;
    }
}   // ---------------------------------------- End of Procedure LongUSTable.

// ==========================================================================
void ShortUSTable(void)
{
  double altKm;
  double sigma,delta,theta;
  double temp,pressure,density, asound;
  double viscosity,kinematicViscosity,vratio;

  std::ofstream Itxt("us2cpp.prt");
  Itxt << " alt  sigma  delta  theta";
  Itxt << "  temp  press    dens     a    visc   k.visc  ratio\n";
  Itxt << " Kft                     ";
  Itxt << "  degR   psf   s/cu.ft   fps  s/ft-s  sq.ft/s  1/ft\n";
  Itxt.setf(std::ios::showpoint | std::ios::uppercase);

  for(int i=-1;i<=65;i++)
    {
      altKm=i*FT2METERS;
      SimpleAtmosphere(altKm, sigma,delta,theta);
      Itxt.setf(std::ios::fixed, std::ios::floatfield);
      Itxt.width(4); Itxt << i;
      Itxt.width(7); Itxt.precision(4); Itxt << sigma;
      Itxt.width(7); Itxt << delta;
      Itxt.width(7); Itxt << theta;

      temp=(KELVIN2RANKINE*TZERO)*theta;
      pressure=(PZERO/PSF2NSM)*delta;
      density=(RHOZERO/SCF2KCM)*sigma;
      asound=(AZERO/FT2METERS)*sqrt(theta);
      Itxt.width(6); Itxt.precision(1); Itxt << temp;
      Itxt.width(7); Itxt.precision(1); Itxt << pressure;
      Itxt.width(10); Itxt.precision(7); Itxt << density;
      Itxt.width(7); Itxt.precision(1); Itxt << asound;

      viscosity=(1.0/PSF2NSM)*MetricViscosity(theta);
      kinematicViscosity=viscosity/density;
      vratio=asound/kinematicViscosity;
      Itxt.width(6); Itxt.precision(3); Itxt << 1.0E6*viscosity;
      Itxt.setf(std::ios::scientific, std::ios::floatfield);
      Itxt.width(11); Itxt.precision(2); Itxt << kinematicViscosity;
      Itxt.setf(std::ios::fixed, std::ios::floatfield);
      Itxt.width(5); Itxt.precision(2); Itxt << 1.0E-6*vratio << std::endl;

    }
}   // --------------------------------------- End of Procedure ShortUSTable.

// ==========================================================================
void LongSITable()
{
  double altKm;
  double sigma,delta,theta;
  double temp,pressure,density,asound;
  double viscosity,kinematicViscosity;

  std::ofstream Itxt("si1cpp.prt");
  Itxt << "units of alt are Km; pressure N/sq.m.; density is kg/cu.m.\n";
  Itxt << "speed of sound is m/s; viscosity is kg per m-s;\n";
  Itxt << " alt    sigma       delta     theta ";
  Itxt << " temp   press     dens      a   visc   k.visc" << std::endl;
//  Itxt << "kinematic viscosity is square meters per second.\n";
//  Itxt << "  Km                              ";
//  Itxt << "   K    N/sq.m   kg/cu.m  m/sec kg/m-s sq.m/s\n";
  Itxt.setf(std::ios::showpoint | std::ios::uppercase | std::ios::floatfield);

  for(int i=-1; i<=43; i++)
    {
      altKm=2*i;
      Atmosphere(altKm, sigma,delta,theta);
      Itxt.width(4); Itxt << 2*i;
      Itxt.setf(std::ios::scientific, std::ios::floatfield);
      Itxt.width(12); Itxt.precision(4); Itxt << sigma;
      Itxt.width(12); Itxt << delta;
      Itxt.setf(std::ios::fixed, std::ios::floatfield);
      Itxt.width(7); Itxt.precision(4); Itxt << theta;

      temp=TZERO*theta;
      pressure=PZERO*delta;
      density=RHOZERO*sigma;
      asound=AZERO*sqrt(theta);
      Itxt.width(7); Itxt.precision(1); Itxt << temp;
      Itxt.setf(std::ios::scientific, std::ios::floatfield);
      Itxt.width(10); Itxt.precision(3); Itxt << pressure;
      Itxt.width(10); Itxt.precision(3); Itxt << density;
      Itxt.setf(std::ios::fixed, std::ios::floatfield);
      Itxt.width(6); Itxt.precision(1); Itxt << asound;

      viscosity=MetricViscosity(theta);
      kinematicViscosity=viscosity/density;
      Itxt.width(6); Itxt.precision(2); Itxt << 1.0E6*viscosity;
      Itxt.setf(std::ios::scientific, std::ios::floatfield);
      Itxt.width(10); Itxt.precision(2); Itxt << kinematicViscosity << std::endl;
    }
}   // ---------------------------------------- End of Procedure LongSITable.

// ==========================================================================
void ShortSITable()
{
  double altKm;
  double sigma,delta,theta;
  double temp,pressure,density,asound;
  double viscosity,kinematicViscosity,vratio;

  std::ofstream Itxt("si2cpp.prt");
  Itxt << " alt  sigma  delta  theta";
  Itxt << "  temp  press  dens   a    visc   k.visc ratio\n";
  Itxt << "  Km                     ";
  Itxt << "    K  N/sq.m  kcm  m/sec kg/m-s  sq.m/s  1/m\n";
  Itxt.setf(std::ios::showpoint | std::ios::uppercase);

  for(int i=-1; i<=40; i++)
    {
      altKm=0.5*i;
      SimpleAtmosphere(altKm, sigma,delta,theta);
      Itxt.setf(std::ios::fixed, std::ios::floatfield);
      Itxt.width(4); Itxt.precision(1); Itxt << altKm;
      Itxt.width(7); Itxt.precision(4); Itxt << sigma;
      Itxt.width(7); Itxt << delta;
      Itxt.width(7); Itxt << theta;

      temp=TZERO*theta;
      pressure=PZERO*delta;
      density=RHOZERO*sigma;
      asound=AZERO*sqrt(theta);
      Itxt.width(6); Itxt.precision(1); Itxt << temp;
      Itxt.width(7); Itxt << long(pressure);              // print as integer
      Itxt.width(6); Itxt.precision(3); Itxt << density;
      Itxt.setf(std::ios::fixed, std::ios::floatfield);
      Itxt.width(6); Itxt.precision(1); Itxt << asound;

      viscosity=MetricViscosity(theta);
      kinematicViscosity=viscosity/density;
      vratio=asound/kinematicViscosity;
      Itxt.setf(std::ios::fixed, std::ios::floatfield);
      Itxt.width(6); Itxt.precision(2); Itxt << 1.0E6*viscosity;
      Itxt.setf(std::ios::scientific, std::ios::floatfield);
      Itxt.width(10); Itxt.precision(2); Itxt << kinematicViscosity;
      Itxt.setf(std::ios::fixed, std::ios::floatfield);
      Itxt.width(6); Itxt.precision(2); Itxt << 1.0E-6*vratio << std::endl;
   }
}   // --------------------------------------- End of Procedure ShortSITable.

// ==========================================================================
void Atmosphere(const double  alt,                  // geometric altitude, km.
		double& sigma,           // density/sea-level standard density
		double& delta,         // pressure/sea-level standard pressure
		double& theta)   // temperature/sea-level standard temperature
// Compute the temperature,density, and pressure in the standard atmosphere
// Correct to 86 km.  Only approximate thereafter.
{
  const double REARTH=6369.0;    // radius of the Earth (km)
  const double GMR = 34.163195;
  const int NTAB = 8;
  int i,j,k;

  static double htab[NTAB] = {0.0,  11.0, 20.0, 32.0, 47.0,
			     51.0, 71.0, 84.852 };
  static double ttab[NTAB] = { 288.15, 216.65, 216.65, 228.65, 270.65,
			      270.65, 214.65, 186.946 };
  static double ptab[NTAB] = { 1.0, 2.2336110E-1, 5.4032950E-2, 8.5666784E-3,
     1.0945601E-3, 6.6063531E-4, 3.9046834E-5, 3.68501E-6 };
  static double gtab[NTAB] = { -6.5, 0, 1.0, 2.8, 0, -2.8, -2.0, 0 };

  double h=alt*REARTH/(alt+REARTH);     //  geometric to geopotential altitude

  i=0; j=NTAB-1;  // starting values for binary search
  do
    {
      k=(i+j)/2;
      if (h < htab[k]) j=k; else i=k;
    }  while (j > i+1);

  double tgrad=gtab[i];                      // temp. gradient of local layer
  double tbase=ttab[i];                      // base temp. of local layer
  double deltah=h-htab[i];                   // height above local base
  double tlocal=tbase+tgrad*deltah;          // local temperature
  theta=tlocal/ttab[0];                                  // temperature ratio

  if (0.0 == tgrad)                                         // pressure ratio
    delta=ptab[i]*exp(-GMR*deltah/tbase);
  else
    delta=ptab[i]*pow(tbase/tlocal, GMR/tgrad);

  sigma=delta/theta;                                        //  density ratio
}   // ------------------------------------------- End of function Atmosphere

// ==========================================================================
void SimpleAtmosphere(
	const double alt,                           // geometric altitude, km.
	double& sigma,                   // density/sea-level standard density
	double& delta,                 // pressure/sea-level standard pressure
	double& theta)           // temperature/sea-level standard temperature

// Compute the temperature,density, and pressure in the standard atmosphere
// Correct to 20 km.  Only approximate thereafter.
{
  const double REARTH = 6369.0;    // radius of the Earth (km)
  const double GMR    = 34.163195;   // gas constant


  double h=alt*REARTH/(alt+REARTH);     //  geometric to geopotential altitude

  if (h<11.0)
    {                                                          // Troposphere
      theta=(288.15-6.5*h)/288.15;
      delta=pow(theta, GMR/6.5);
    }
  else
    {                                                         // Stratosphere
      theta=216.65/288.15;
      delta=0.2233611*exp(-GMR*(h-11.0)/216.65);
    }

  sigma=delta/theta;
}   // ------------------------------------- End of function SimpleAtmosphere

// ==========================================================================
double MetricViscosity(const double theta)
{
  const double TZERO = 288.15;               // sea level temperature, kelvins
  const double BETAVISC = 1.458E-6;     // constant, N-sec/(sq.m-sqrt(kelvins)
  const double SUTH = 110.4;                 // Sutherland's constant, kelvins

  double t=theta*TZERO;                              // temperature in kelvins
  return BETAVISC*sqrt(t*t*t)/(t+SUTH);            // viscosity in kg/(m-sec)
}   // -------------------------------------- End of function MetricViscosity
